core/offchain.go

```diff
@@ -187,3 +214,20 @@ func (bc *BlockChain) CommitOffChainData(
 
 	return CanonStatTy, nil
 }
+
+func (bc *BlockChain) removeInactiveSigners(
+	missedThreshold map[common.Address]struct{},
+) error {
+	addrs, err := bc.ReadValidatorList()
+	if err != nil {
+		return err
+	}
+	eligible := []common.Address{}
+	for _, addr := range addrs {
+		if _, mia := missedThreshold[addr]; !mia {
+			eligible = append(eligible, addr)
+		}
+	}
+
+	return bc.WriteValidatorList(bc.db, eligible)
```

Comment:
If we remove those validators from the list, then their is no wait to iterate/list all validators anymore. Then the explorer won't be able to show them. Why do we need to change the ValidatorList? Can we do filtering instead on reading ValidatorList?

Reply:
because median stake is reading from this (as well as committee asssignment), the inactive ones are coming from the block and I remove the inactive ones from the list.

Reply:
if I do filtering, then it means you want me to read from state each time any median stake computation happens, either via RPC or commitee assignment. Then you defeat the purpose of this cache in the first place.